scripts/dss_v2_apdapter.py

```diff
@@ -0,0 +1,489 @@
+        # files before starting the upload of this bundle.
+        schema_types = {}  # Mapping of file uuid to schema type
+        manifest_entries: MutableJSON = {}  # Mapping of file names to manifest entry
+        file_name_re = re.compile(r'^(\w+)_\d+\.json$')  # ex. 'cell_suspension_0.json'
+        t1 = time.perf_counter()
+        manifest_entry: MutableJSON
+        for manifest_entry in bundle['files']:
+            manifest_entries[manifest_entry['name']] = manifest_entry
+            if manifest_entry['name'] == 'project_0.json':
+                schema_types[manifest_entry['uuid']] = 'project'
+            elif manifest_entry['name'] == 'links.json':
+                schema_types[manifest_entry['uuid']] = 'links'
+            elif manifest_entry['indexed']:  # Metadata files
+                # FIXME: get schema type from 'describeBy' https://github.com/databiosphere/azul/issues/1831
+                match = file_name_re.match(manifest_entry['name'])
+                if match:
+                    schema_types[manifest_entry['uuid']] = match.group(1)
+                else:
+                    self.log_warning(bundle_fqid,
+                                     manifest_entry['name'],
+                                     'Indexed file has unknown file name format. Bundle skipped.')
+                    return
+            else:  # Data files
+                schema_types[manifest_entry['uuid']] = 'data'
+                manifest_entries[manifest_entry['name']]['new_name'] = self.data_file_new_name(
+                    manifest_entry['name'],
+                    manifest_entry['uuid'],
+                    manifest_entry['version']
+                )
+        missing_files = []
+        if 'project_0.json' not in manifest_entries:
+            missing_files.append('project_0.json')
+        if 'links.json' not in manifest_entries:
+            missing_files.append('links.json')
+        if missing_files:
+            self.log_warning(bundle_fqid, '', f"No {' or '.join(missing_files)} found in bundle. Bundle skipped.")
+            return
+        if self.args.debug:
+            benchmarks.append((time.perf_counter() - t1, 'to build data file name mapping'))
+
+        # Build the new links.json before we start uploading any parts of the bundle
+        blob_key = self._build_blob_key(manifest_entries['links.json'])
+        old_links_json = json.loads(self._download_blob_as_string(blob_key))
+        project_uuid = manifest_entries['project_0.json']['uuid']
+        supp_files = [uuid for uuid, schema in schema_types.items() if schema == 'supplementary_file']
+        try:
+            new_links_json = self._build_new_links_json(old_links_json, schema_types, project_uuid, supp_files)
+        except Exception as e:
+            self.log_warning(bundle_fqid, 'links.json', str(e))
+            return
+
+        # Now loop over the manifest entries to upload or copy files to the
```

Comment:
Low priority: This function is now pretty spaghetti. Any chance to break it up into separate functions?

That would make it more explicit how state flows between. If losing the benchmarks code would make the signatures simpler, I'd be ok with losing it. Not a big fan anyways.

Another potential refactoring is to introduce a class of which there is an instance per bundle.